src/v/model/token_bucket.cc

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "model/token_bucket.h"
+
+#include "seastarx.h"
+#include "ssx/semaphore.h"
+#include "vassert.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/later.hh>
+
+throttler::throttler(
+  size_t rate, ss::sstring&& name, size_t burst_rate_multiplier)
+  : _rate(rate)
+  , _burst_rate_multiplier(burst_rate_multiplier)
+  , _sem{rate, std::move(name)}
+  , _last_refresh(clock_type::now())
+  , _refresh_timer([this] { handle_refresh(); }) {}
+
+ss::future<> throttler::throttle(size_t size, ss::abort_source& as) {
+    _refresh_timer.cancel();
+    refresh();
+
+    /*
+     * when try_wait succeeds it implies that there are no waiters so there
+     * is no risk in returning without arming the refresh timer.
+     */
+    if (_sem.try_wait(size)) {
+        return ss::now();
+    }
+
+    auto elapsed = clock_type::now() - _last_refresh;
+    if (elapsed >= refresh_interval) {
+        _refresh_timer.arm(refresh_interval);
+    } else {
+        _refresh_timer.arm(refresh_interval - elapsed);
+    }
+
+    return _sem.wait(as, size);
+}
+
+ss::future<bool> throttler::try_throttle(size_t size) {
```

Comment:
this should not return future as it is synchronous method

Reply:
fixed